### PR TITLE
feat(utils): add more utils for testing contracts

### PR DESCRIPTION
**Pull request by martinkaintas**
Adds utils mentioned in #439
Adds `emittedEvents` util

### CODE REVIEW

**Comment by thepiwo** [comment]
TBH I disagree with this approach, this is exactly what we had with the old version of aeproject, then growing out of maintainability. These utils if we want them should be part of the sdk, aeproject users should use as plain sdk/chai as possible.

**Comment by marc0olo** [comment]
@thepiwo with my proposal or with the PR in general? these helpers use plain chai, don't they? not sure if I understand your comment.

puh, no idea. I guess we have to generally take a step back and actually define how our desired tooling landscape should look like. otherwise we will always end up in situations like this. maybe sth. the foundation should take over now. I think we will end up stuck if this isn't tackled the right way. cc @dincho @benhus 

personally I can only say that I want to use certain things and that I need specific functionality.

**Comment by thepiwo** [comment]
with both, I don't think its needed.

The problem is, if either the sdk changes or chai is updated aeproject needs maintenance in this case (for more and more functions). So if we start adding many helpers here maintenance effort grows. So I would say, if the user can do it with chai or the sdk in his code he should do it there or in some separate from aeproject library (e.g. the sdk).

This approach was at least the discussion to my re-write and continued maintainance of the current version of aeproject. I am not the decider here, but in the past it has been developed with many features by an external company, then also neglected by them and left with to much maintainance effort due to lots of code. If the decision is made to go back to this approach I am not keen to re-write it again after the next phase of neglection.

**Comment by thepiwo** [comment]
> personally I can only say that I want to use certain things and that I need specific functionality.

as you see this code is very few lines, what keeps you from having this specific logic either added to sdk if generally used in AE or specified in your own project on a case by case basis?

**Comment by marc0olo** [comment]
> as you see this code is very few lines, what keeps you from having this specific logic either added to sdk if generally used in AE or specified in your own project on a case by case basis?

I think it can also be in a separate library. doesn't have to be aeproject. but I don't think it's core functionality that should be included in the SDK.

I am not sure about the best place to include these helpers and I also don't wanna decide this (alone). I think devs shouldn't have to copy/paste this in every project again and again.

so assuming SDK is the wrong place and we don't want to include it here, would we just provide a snippet in the docs which users can copy paste? (we'd even have to maintain this then 🤔 🤣 )

**Comment by thepiwo** [comment]
> shouldn't have to copy/paste this in every project again and again

haven't seen it copy pasted in different projects yet, verbosity of error handling is also developer preference

> core functionality that should be included in the SDK

for the none chai topics, why not?

> a snippet in the docs

different answer depending on snippet, for events handling I think it should be part of the sdk, for chai usage I think thats up to the developer to decide how to solve, there are also chai libraries to improve async error handling over chai defaults as far as I know.

**Comment by marc0olo** [comment]
alright. the question is how to proceed. I guess we can close this PR then.

can @martinkaintas help in any way here and add at least event handling helpers e.g. to the SDK if we agree it should be there? cc @davidyuk @loxs @dincho

File: docs/lib.md [hunk 1]
```diff
@@ -48,6 +48,21 @@ utils.getDefaultAccounts();
 ```
 Get the pre-funded default accounts as `MemoryAccount`, so they can be used natively using the `aeSdk`.
 
+```javascript
+utils.getUsedFee(() => contract.methods.X(...));
+```
+Calls X and returns the total fee paid for the transaction
+
+```javascript
+utils.failsWith(() => contract.methods.X(...), 'errorMsg');
```
Comment (davidyuk) [inline docs/lib.md:57]: Isn't it `await expect(contract.methods.X(...)).to.be.rejectedWith(ErrorClass, 'errorMsg')`?
https://www.chaijs.com/plugins/chai-as-promised/

File: src/lib/utils.js [hunk 2]
```diff
@@ -114,6 +115,48 @@ const rollbackSnapshot = async (aeSdk) => {
...
+ * @param {Function} f  function to be called
+ * @returns {Promise<BigInt>} the total fee that was paid for the transaction
+ */
+const getUsedFee = async (f) => {
+  const {
+    result: { gasPrice, gasUsed },
+    txData: { tx: { fee } },
+  } = await f();
+  // eslint-disable-next-line no-undef
+  return gasPrice * BigInt(gasUsed) + fee;
```
Comment (davidyuk) [inline src/lib/utils.js:130]: I agree that this should be on sdk side

File: docs/lib.md [hunk 3]
```diff
@@ -48,6 +48,21 @@ utils.getDefaultAccounts();
...
+```
+Calls X and checks if it fails with expected error
+
+```javascript
+utils.emittedEvents(() => contract.methods.X(...), ['eventName1']);
```
Comment (davidyuk) [inline docs/lib.md:62]:
```suggestion
const { decodedResult, decodedEvents } = await contract.methods.X(...);
expect(decodedEvents.map(e => e.name)).to.have.members(['eventName1']);
```
Can something like this be used instead?

**Comment by thepiwo** [comment]
closing this for now, feel free to reopen if necessary

**Pull request closed without merging**